### DISCUSSION TITLE
if not FileTruststoreProviderFactory Which class is invoking the truststore, and where can I find the truststore logs for the same?

### DISCUSSION THREAD

**Discussion by chetanrun** [keycloak/keycloak#36205] [Q&A]
### Description

KC version: 21.0.1

The FileTruststoreProviderFactory class is not being used for loading the truststore because:

Problem: I wanted to verify which certificates were loaded during and after startup. To do this, I added logging to the FileTruststoreProviderFactory class methods:

create(session) -- is never called during keycloak statup
init(config)
postinit(factory)

The goal was to check if the TruststoreProvider was correctly set. However, the provider instance is always null in these methods. Here's the snippet I used:

TruststoreProvider provider_ = TruststoreProviderSingleton.get();
log.info("-------------------provider: " + provider_);

Docker Compose Settings:

- KC_HTTPS_TRUST_STORE_FILE=/opt/keycloak/bin/TrustStore.jks
- KC_HTTPS_TRUST_STORE_TYPE=jks
- KC_HTTPS_TRUST_STORE_PASSWORD=**********
- KC_HTTPS_CLIENT_AUTH=request

Will there be any runtime resolution apart FileTruststoreProviderFactory?

Observation: During certificate-based authentication, the browser lists supported certificates, and upon selection, I can successfully log in using a certificate. But how?

And if we use deprecated config as below, then the FileTruststoreProviderFactory is rightly called and the truststoreprovider is prepared as it can be seen from the log file but the **observation**: During certificate-based authentication, the browser doesn't list supported certificates, automatically failing login. 

      - KC_SPI_TRUSTSTORE_FILE_FILE=/opt/keycloak/bin/TrustStore.jks
      - KC_SPI_TRUSTSTORE_FILE_PASSWORD=******
      - KC_SPI_TRUSTSTORE_FILE_TYPE=jks

```[tasklist]
### Tasks
```

**Comment by shawkins** [comment]
> Observation: During certificate-based authentication, the browser lists supported certificates, and upon selection, I can successfully log in using a certificate. But how?

With your HTTPS settings, that truststore will be used - those are directly mapped to quarkus options.

Note that in KC 21 the FileTruststoreProviderFactory does nothing if the store path, password, and verification policy are not set.

> And if we use deprecated config as below, then the FileTruststoreProviderFactory is rightly called and the truststoreprovider is prepared as it can be seen from the log file but the observation: During certificate-based authentication, the browser doesn't list supported certificates, automatically failing login.

Until KC 24 with https://github.com/keycloak/keycloak/issues/24148 the file truststore provider and the https truststore were unrelated. Setting the file truststore provider in kc 24+ will effectively set up the system truststore, which may then be used by default by the https listener logic in quarkus.
